18F/identity-idp
# LG-12294: Send single aggregated email notification for new device sign-in

## 🎫 Ticket

[LG-12294](https://cm-jira.usa.gov/browse/LG-12294)

## 🛠 Summary of changes

Implements the dispatching of a new aggregated email notification for sign-in, which happens 5 minutes after successful email and password submission, or immediately upon successful MFA, whichever occurs first.

This builds upon ("wires up") previous work in #10314 (email template) and #10317 (worker for delayed notification).

## 📜 Testing Plan

Verify that you are sent an email immediately upon full authentication:

1. In a private browsing session ("Incognito"), visit http://localhost:3000
2. Sign in with email and password
3. Complete MFA step
4. Observe you see an email notification immediately with listed events
   - Note that it's likely to show "unknown location" since events save `'::1'` as the IP address in local development

Verify that you are sent an email after delay upon partial authentication:

1. Optional: Change configured delay and/or worker frequency to avoid having to wait a full 5 minutes:
   - Set `new_device_alert_delay_in_minutes: 1` in `config/application.yml` (changes notification to send after 1 minute instead of after 5 minutes)
   - Set [`cron_5m` in `job_configurations.rb`](https://github.com/18F/identity-idp/blob/0e68f0919d9168f43a5c18fa9b5c355501212ad5/config/initializers/job_configurations.rb#L3) to `'0/1 * * * *'` (changes jobs to run every 1 minute instead of every 5 minutes)
2. In a private browsing session ("Incognito"), visit http://localhost:3000
3. Sign in with email and password
4. Wait until the next worker run after the delay specified in configuration
5. Observe you see an email notification with listed events
   - Note that it's likely to show "unknown location" since events save `'::1'` as the IP address in local development

## 👀 Screenshots

Partial Authentication|Full Authentication
---|---
![image](https://github.com/18F/identity-idp/assets/1779930/6652aa6b-e657-46b6-af4b-0754c7b30169)|![image](https://github.com/18F/identity-idp/assets/1779930/25e55e7e-6b09-47c1-bc09-0e417951fdcf)
